File: src/main/scala/demos/fp/order/anemic/OrderService.scala
package demos.fp.order.anemic

object OrderService:
    def addToOrder(order: Order, productId: String, amount: Int): Order =
        val existingOrderLineAndIndex =
            order.lines.zipWithIndex.filter(_._1.productId == productId).headOption
        val updatedLines = existingOrderLineAndIndex match
            case None =>
                val newLine = OrderLineService.createOrderLine(productId, amount)
                order.lines.appended(newLine)
            case Some((existingOrderLine, index)) =>
                val updatedLine = OrderLineService.addToOrderLine(existingOrderLine, amount)
                order.lines.updated(index, updatedLine)
        order.copy(lines = updatedLines)

File: src/main/java/demos/oop/order/anemic/OrderService.java
package demos.oop.order.anemic;

import java.util.List;

class OrderService {

    private Order order;

    public OrderService(Order order) {
        this.order = order;
    }

    public void addToOrder(String productId, int amount) {
        OrderLine existingOrderLine = this.findExistingOrderLine(productId);
        if (existingOrderLine == null) {
            OrderLine newOrderLine = new OrderLine(productId, amount);
            OrderLineService service = new OrderLineService(newOrderLine);
            service.validate();
            this.order.getLines().add(newOrderLine);
        } else {
            OrderLineService service = new OrderLineService(existingOrderLine);
            service.addToOrderLine(amount);
        }
    }

    private OrderLine findExistingOrderLine(String productId) {
        List<OrderLine> lines = this.order.getLines();
        for (OrderLine orderLine : lines) {
            if (orderLine.getProductId().equals(productId)) {
                return orderLine;
            }
        }
        return null;
    }
}
